### PR TITLE
Game engine

### PR DESCRIPTION
**Pull request by LukaszCzubernat**
Game engine - core business logic component that should provide basic game logic, like create or join game, check if ship is hit, decide who is a winner.

### CODE REVIEW

**Comment by YogAzathoth** [comment]
As we were discussing most of the code while it was created therefore I can't actually comment on anything (as we were exchanging comments during the creation process). 
The final renaming is nice, for me it is more descriptive and easier to read and understand what exactly is going and what is responsible for what.
As also I will need some major changes in my part of application to use things there already are created.

File: Util/src/test/java/model/ship/wrapper/util/BattleshipAreaResolverTest.java [hunk 1]
```diff
@@ -15,15 +16,15 @@
...
     @Test
     public void testFindCoordinatesAroundShip(){
         // given
-        Ship smallShip = new Ship(Arrays.asList(new Coordinate(A, 1)));
+        Battleship smallShip = new Battleship(Arrays.asList(new Coordinate(A, 1)));
         List<Coordinate> expectedCoordinatesAroundShip = Arrays.asList(new Coordinate(A, 2), new Coordinate(B, 1), new Coordinate(B, 2));
 
         // when - then
-        assertTrue(ListUtils.isEqualList(ShipAreaResolver.findCoordinatesAroundShip(smallShip), expectedCoordinatesAroundShip));
+        assertTrue(ListUtils.isEqualList(BattleshipAreaResolver.findCoordinatesAroundShip(smallShip), expectedCoordinatesAroundShip));
```
Comment (clinton56) [inline Util/src/test/java/model/ship/wrapper/util/BattleshipAreaResolverTest.java]: :+1:  `org.apache.commons` collections utils:)

File: Core/src/main/java/controller/arbiter/impl/BattleshipArbiter.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+import controller.arbiter.util.ArbiterDecisionEngine;
+import model.coordinate.Coordinate;
+import model.client.Client;
+import model.ship.Ship;
+
+/**
+ * Created by lucz on 22.07.16.
+ */
+public class BattleshipArbiter implements Arbiter {
+    private ArbiterDecisionEngine arbiterDecisionEngine = new ArbiterDecisionEngine();
```
Comment (clinton56) [inline Core/src/main/java/controller/arbiter/impl/BattleshipArbiter.java]: You won't be able to mock and test it correctly by creating it here. I would suggest to try dependency injection.

File: Core/src/main/java/controller/arbiter/util/ArbiterDecisionEngine.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class ArbiterDecisionEngine {
+
+    private Map<Client, List<Coordinate>> clientsCoordinates = new HashMap<>();
+
+    public void addCoordinatesFromShip(Client client, Ship ship) {
+        if (CollectionUtils.isEmpty(clientsCoordinates.get(client)))
```
Comment (clinton56) [inline Core/src/main/java/controller/arbiter/util/ArbiterDecisionEngine.java]: Using braces for `if` block improves readability and makes it easier to add something in the future.

File: Core/src/main/java/controller/cache/impl/BattleshipGameCache.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    @Override
+    public Set<Client> getCachedClients(UUID gameID) {
+        return gamesCache.get(gameID);
+    }
+
+    @Override
+    public Set<UUID> getAvailableGames(){
+        return gamesCache.keySet().stream().filter(key -> gamesCache.get(key).size()<2).collect(Collectors.toSet());
```
Comment (clinton56) [inline Core/src/main/java/controller/cache/impl/BattleshipGameCache.java]: :+1:  stream

File: Core/src/main/java/controller/arbiter/impl/BattleshipArbiter.java (hunk 2 above)
Comment (LIttleAncientForestKami) [inline Core/src/main/java/controller/arbiter/impl/BattleshipArbiter.java]: @clinton56 he will with test-only constructor or `@Mock` annotation.

File: Core/src/main/java/controller/arbiter/Arbiter.java [hunk 5]
```diff
@@ -1,27 +1,18 @@
 package controller.arbiter;
 
 import model.coordinate.Coordinate;
+import model.client.Client;
+import model.ship.Ship;
 
-import java.util.ArrayList;
-import java.util.List;
+/**
+ * Created by lucz on 01.08.16.
```
Comment (LIttleAncientForestKami) [inline Core/src/main/java/controller/arbiter/Arbiter.java]: lucz? For Javadoc on an interface, too little.

File: Core/src/main/java/controller/arbiter/Arbiter.java [hunk 6]
```diff
@@ -1,27 +1,18 @@
...
+ */
+public interface Arbiter {
+    void registerShipForClient(Client client, Ship ship);
 
-public class Arbiter {
+    boolean isHit(Client client, Coordinate hitCoordinate);
 
-    private List<Coordinate> coordinates = new ArrayList<>();
+    boolean isLosing(Client client);
```
Comment (LIttleAncientForestKami) [inline Core/src/main/java/controller/arbiter/Arbiter.java]: Interesting. How can I determine someone "is losing" and what do I do with it?

File: Core/src/main/java/controller/cache/GameCache.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
+package controller.cache;
+
+import model.client.Client;
+
+import java.util.Set;
+import java.util.UUID;
+
+/**
+ * Created by lucz on 04.08.16.
```
Comment (LIttleAncientForestKami) [inline Core/src/main/java/controller/cache/GameCache.java]: Same as previously.

File: Core/src/test/java/controller/arbiter/impl/BattleshipArbiterTest.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+
+    @Test
+    public void testIsFirstClientIsLosing(){
+        // given
+        Coordinate hitCoordinate = new Coordinate(A, 1);
+
+        // when
+        arbiter.isHit(firstClient, hitCoordinate);
```
Comment (LIttleAncientForestKami) [inline Core/src/test/java/controller/arbiter/impl/BattleshipArbiterTest.java]: Ouch! Poor arbiter, getting all hit like that. :-)

File: Core/src/test/java/controller/arbiter/impl/BattleshipArbiterTest.java [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+        // then
+        assertTrue(arbiter.isLosing(firstClient));
```
Comment (LIttleAncientForestKami) [inline Core/src/test/java/controller/arbiter/impl/BattleshipArbiterTest.java]: And even losing!

File: Core/src/test/java/controller/arbiter/impl/BattleshipArbiterTest.java [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    public void testIsSecondClientHitShip() {
+        // given
+        Coordinate hitCoordinate = new Coordinate(A, 3);
+
+        // when - then
+        assertTrue(arbiter.isHit(secondClient, hitCoordinate));
+    }
+
+    @Test
+    public void testIsSecondClientNotHitShip()  {
```
Comment (LIttleAncientForestKami) [inline Core/src/test/java/controller/arbiter/impl/BattleshipArbiterTest.java]: Czy nie jest drugi klient trafiony statek. :confounded:

File: Core/src/test/java/controller/arbiter/util/ArbiterDecisionEngineTest.java [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+    public void addCoordinatesTest() {
+        // Given
+        ArbiterDecisionEngine arbiterDecisionEngine = new ArbiterDecisionEngine();
+        Client player = new Client(UUID.randomUUID());
+
+        // When
+        arbiterDecisionEngine.addCoordinatesFromShip(player, ship);
+
+        // Then
+        assertTrue(arbiterDecisionEngine.contains(player, new Coordinate(A, 1)));
```
Comment (LIttleAncientForestKami) [inline Core/src/test/java/controller/arbiter/util/ArbiterDecisionEngineTest.java]: Soft assertions.

File: Core/src/test/java/controller/arbiter/util/ArbiterDecisionEngineTest.java [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+        ship = new Battleship(Arrays.asList(new Coordinate(A, 1)));
+    }
+
+    @AfterMethod
+    public void cleaning() {
+        ship = null;
+    }
+
+    @Test
+    public void addCoordinatesTest() {
```
Comment (LIttleAncientForestKami) [inline Core/src/test/java/controller/arbiter/util/ArbiterDecisionEngineTest.java]: Would be cool to have this parametrized. It should be, I think.

File: Core/src/test/java/controller/arbiter/util/ArbiterDecisionEngineTest.java [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+        // When
+        arbiterDecisionEngine.addCoordinatesFromShip(player, ship);
+
+        // Then
+        assertTrue(arbiterDecisionEngine.contains(player, new Coordinate(A, 1)));
+        assertFalse(arbiterDecisionEngine.contains(player, new Coordinate(D, 10)));
+    }
+
+    @Test
+    public void winningConditionNotOccursTest() {
```
Comment (LIttleAncientForestKami) [inline Core/src/test/java/controller/arbiter/util/ArbiterDecisionEngineTest.java]: `testWinningConditionNotHappening`? `noWinYetTest`? There must be something better.

**Comment by LukaszCzubernat** [comment]
Is it good enough or not?

**Comment by clinton56** [comment]
:+1:  4 adding reporting and JavaDocs. I like your PR because of Apache commons, Java 8 and the number of tests. I accept the solution. :ghost:

**Pull request merged**